### PR TITLE
SCRD-2756 post-deployment sanity checks

### PR DESCRIPTION
**Pull request by aspiers**
See [SCRD-2756 (Add CI checks to ensure SLES updates can be installed any time)](https://suse-jira.dyndns.org/browse/SCRD-2756)

**NOTE:** hasn't been successfully tested the whole way through in a single run yet, due to unrelated issues preventing everything else working.

Check that Ardana doesn't mess with the filesystem inappropriately by removing or modifying rpm-owned files, or adding files into `/usr`.  This is done by running `rpm -Va` and `rpm -qf` before and after Ardana deployment, and comparing the results.

If any tampering is discovered, `init.yml` will fail with a helpful error.

The checks will run even if `site.yml` or tempest failed.  A list of all failures (including `site.yml` and tempest) is displayed at the end, to avoid hiding multiple failures behind one failure.

### CODE REVIEW

File: scripts/jenkins/ardana/ansible/post-deploy-checks.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  failed_when: false
+  register: rpm_verify_delta
+
+- name: Obtain list of rpm-owned files removed by deployment
+  shell: echo "{{ rpm_verify_delta.stdout }}" | grep '^missing '
+  failed_when: false
+  register: rpm_verify_removed_files
+
+- name: Obtain list of rpm-owned non-config files modified by deployment
+  shell: echo "{{ rpm_verify_delta.stdout }}" | egrep -v '^(missing |.........  c)'
```
Comment (rhafer) [inline scripts/jenkins/ardana/ansible/post-deploy-checks.yml]: You might want to ignore ghost files as well. They should show up with a 'g'  in the rpm -Va output. (It might also be an option to call rpm with `--noghost`)

Comment (aspiers) [reply scripts/jenkins/ardana/ansible/post-deploy-checks.yml]: Good catch, thanks!!

Comment (aspiers) [reply scripts/jenkins/ardana/ansible/post-deploy-checks.yml]: @rhafer Done, although I'm still blocked from testing due to [bsc#1081005 (Ardana horizon venv fails during the installation)](https://bugzilla.suse.com/show_bug.cgi?id=1081005)

File: scripts/jenkins/ardana/ansible/post-deploy-checks.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+- name: Take snapshot of modified/deleted rpm-owned files post deployment
+  shell: rpm -Va | grep -vxFf "{{ pre_deploy_rpm_verification }}"
```
Comment (cmurphy) [inline scripts/jenkins/ardana/ansible/post-deploy-checks.yml]: I think you might need `become: true` for this, at least on my system I see a lot of "missing [...] (Permission denied)" when I run this without sudo.

Comment (aplanas) [reply scripts/jenkins/ardana/ansible/post-deploy-checks.yml]: @cmurphy is right, this can produce false `missing c` files in the report, just because rpm cannot read the directory

Comment (aspiers) [reply scripts/jenkins/ardana/ansible/post-deploy-checks.yml]: Good catch thanks, that's the bit I was unable to test properly yet due to the Horizon issue. Fixed now.

**Comment by aspiers** [comment]
Running new test of this, hopefully it will finally work: https://ci.suse.de/job/ardana-job-aspiers/16/console

**Comment by aspiers** [comment]
Done several more tests and ironed out some further issues - vaguely hopeful that https://ci.suse.de/job/ardana-job-aspiers/21/console is the first one which will work.

**Comment by aspiers** [comment]
@aplanas I addressed your concern so your review is out of date now.

**Comment by aspiers** [comment]
Success! https://ci.suse.de/job/ardana-job-aspiers/21/consoleText shows useful output, however we cannot merge this yet because of the failures.

**Comment by aspiers** [comment]
I have now added whitelists to this, so in theory it should not add any new failures to `ardana-job`.  Needs a re-review though, and of course a successful `ardana-job` test run.

**Comment by sjamgade** [comment]
this might be a bit late, but have you thought of using a single shell script ? there are just a few use cases I saw where a module really helped

**Comment by aspiers** [comment]
Yes I did think of that, but not sure that would help much, and anyway yes it's too late now.  Not sure what you mean by a module here?

**Comment by sjamgade** [comment]
Just bouncing off another idea, there are [ansible callback plugins](https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/callback), then this verification could done on [task start](https://github.com/ansible/ansible/blob/508a9c20dcf27bf118f7e9528a9c77e71ed29312/lib/ansible/plugins/callback/__init__.py#L285) or [play start](https://github.com/ansible/ansible/blob/508a9c20dcf27bf118f7e9528a9c77e71ed29312/lib/ansible/plugins/callback/__init__.py#L300), and enabling/disabling  plugin is just a configuration change

**Comment by aspiers** [comment]
Rebased and trying a re-run: [ardana-job#684 (/deployerincloud-lite)](https://ci.suse.de/job/ardana-job/684/console)

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# alternatives system is expected to change stuff under /usr
+-> /etc/alternatives/.
+
+# Installed by osconfig/ansible/role/udev/tasks/configure.yml
+/usr/lib/systemd/system/ardana-udev\.service
+
+# FIXME: where does this come from?
```
Comment (toabctl) [inline scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: From http://git.suse.provo.cloud/cgit/ardana/ardana-ansible/tree/roles/installui/defaults/main.yml#n38 (used in one of the tasks)

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+# Created when python-selinux rpm is installed
+/usr/lib64/python2.7/site-packages/selinux/__init__.pyc
+
+# Ardana installs stuff in here, e.g. links to /opt/stack/service/.../venv/bin/...
+/usr/local/bin/.
+
+/usr/sbin/httpd -> /usr/sbin/httpd-prefork
+
+# Remove when https://gerrit.suse.provo.cloud/#/c/3336/ is merged
```
Comment (toabctl) [inline scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: Is merged

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt (hunk 3 above)
Comment (aspiers) [reply scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: Thanks!

Comment (aspiers) [reply scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: This has since been moved to `/etc` but I think it should be moved to the rpm: https://bugzilla.suse.com/show_bug.cgi?id=1082708#c4

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt (hunk 4 above)
Comment (aspiers) [reply scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: Thanks, removing.

**Comment by aspiers** [comment]
@toabctl Documentation added.  Note that this still shouldn't be merged until we get a successful run :-(

**Comment by aspiers** [comment]
@JanZerebecki Thanks. Yes I'm not too worried about those corner cases.

**Comment by aspiers** [comment]
Revamped to run the tests even if `site.yml` or tempest fails.  Testing again via https://ci.suse.de/job/ardana-job/875/consoleText

**Comment by aspiers** [comment]
Success!

**Comment by aspiers** [comment]
This is ready for re-review - please take a look.

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# alternatives system is expected to change stuff under /usr
+-> /etc/alternatives/.
```
Comment (aplanas) [inline scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: Can you document the use of `->`

File: docs/ardana/rpm-file-checks.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+[`ardana-job`](testing.md) performs some automated checks to ensure
+that Ardana doesn't mess with the filesystem inappropriately by
+removing or modifying rpm-owned files, or adding files into /usr.
+This is done by running `rpm -Va` and `rpm -qf` before and after
+Ardana deployment (including a tempest run, if deployment succeeded),
+and comparing the results.  If any tampering is discovered, `init.yml`
+will fail with a helpful error.
+
+[Whitelists](../../scripts/jenkins/ardana/ansible/files/) are used to
+ignore known exceptions, some of which need to be fixed in the future.
```
Comment (aplanas) [inline docs/ardana/rpm-file-checks.md]: We have more files, like rpm-verify-modification-fixmes.txt, is this also a whitelist? do we need to document this and how is maintained?

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+# Installed by osconfig/ansible/role/udev/tasks/configure.yml
+/usr/lib/systemd/system/ardana-udev\.service
```
Comment (aplanas) [inline scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: do we need to escape? so this is some regex file?

File: scripts/jenkins/ardana/ansible/init.yml [hunk 8]
```diff
@@ -252,3 +266,28 @@
...
+    register: tempest_run_result
+    ignore_errors: true
+    when: site_result|succeeded
+
+  - name: Record any failure of tempest
+    set_fact:
+      failures: "{{ failures + ['tempest'] }}"
+    when: tempest_run_result|failed
+
+  - include: post-deploy-checks.yml
```
Comment (aplanas) [inline scripts/jenkins/ardana/ansible/init.yml]: I am wondering if we want to run the post-deploy-checks after site.yml, not after tempest. If the playbooks that install tempest are doing something ugly, maybe is not so wrong.

File: scripts/jenkins/ardana/ansible/pre-deploy-checks.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+- name: Take snapshot of files not owned by a package prior to deployment
+  become: yes
+  failed_when: false
+  shell: |
+    find /usr -print0 | xargs -0 rpm -qf | grep "is not owned" \
+      > "{{ pre_deploy_unowned_files }}"
+
+- name: Take snapshot of modified/deleted rpm-owned files prior to deployment
+  become: yes
+  shell: rpm -Va > "{{ pre_deploy_rpm_verification }}"
```
Comment (aplanas) [inline scripts/jenkins/ardana/ansible/pre-deploy-checks.yml]: I am also wondering that this task is only executed in the deployer, so we will miss all the wrongdoing in the control plane, compute nodes and storage.

Maybe we want to gather all this information (and the rest) in for host: * via some other playbook or role

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt (hunk 7 above)
Comment (aspiers) [reply scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: Yes, it's used by `grep`.

File: scripts/jenkins/ardana/ansible/pre-deploy-checks.yml (hunk 9 above)
Comment (aspiers) [reply scripts/jenkins/ardana/ansible/pre-deploy-checks.yml]: Very good point, thanks!  Urgh, that will require running the checks from the deployer though, since `ardana-ci-launcher` doesn't have access to the other nodes. Let's deal with that in a follow-up so as not to delay the immediate benefit this provides.

File: scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt (hunk 5 above)
Comment (aspiers) [reply scripts/jenkins/ardana/ansible/files/rpm-added-files-whitelist.txt]: Done.

File: docs/ardana/rpm-file-checks.md (hunk 6 above)
Comment (aspiers) [reply docs/ardana/rpm-file-checks.md]: Renamed to `rpm-verify-modification-whitelist.txt` and documented.

**Comment by aspiers** [comment]
@aplanas Thanks for the helpful feedback - I think I addressed all your points now.

**Comment by aspiers** [comment]
@stefannica Sorry - pushed two more minor changes.  I didn't squash them [as per my own advice](https://blog.adamspiers.org/2017/08/16/squash-merging-and-other-problems-with-github/) because that makes reviewing harder, and in this case I don't think it pollutes the history badly enough to justify the extra review work.  Hey, rules are made to be broken ;-)
